exporter/exportertest/mock_receiver.go

```diff
@@ -0,0 +1,237 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+package exportertest // import "go.opentelemetry.io/collector/exporter/exportertest"
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+var errNonPermanent = status.Error(codes.DeadlineExceeded, "non Permanent error")
+var errPermanent = status.Error(codes.Internal, "Permanent error")
+
+type DecisionFunc func() error
+
+type MockReceiver interface {
+	Start() error
+	Stop() error
+	RequestCounter() RequestCounter
+}
+
+type MockReceiverFactory func(DecisionFunc) MockReceiver
+
+// // randomNonPermanentErrorConsumeDecision is a decision function that succeeds approximately
+// // half of the time and fails with a non-permanent error the rest of the time.
+func randomNonPermanentErrorConsumeDecision() error {
+	if rand.Float32() < 0.5 {
+		return errNonPermanent
+	}
+	return nil
+}
+
+// randomPermanentErrorConsumeDecision is a decision function that succeeds approximately
+// half of the time and fails with a permanent error the rest of the time.
+func randomPermanentErrorConsumeDecision() error {
+	if rand.Float32() < 0.5 {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	return nil
+}
+
+// randomErrorsConsumeDecision is a decision function that succeeds approximately
+// a third of the time, fails with a permanent error the third of the time and fails with
+// a non-permanent error the rest of the time.
+func randomErrorsConsumeDecision() error {
+	r := rand.Float64()
+	third := 1.0 / 3.0
+	if r < third {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	if r < 2*third {
+		return errNonPermanent
+	}
+	return nil
+}
+
+type BaseMockConsumer interface {
```

Comment:
The interface is not being used. I think we should use it instead of exposing the struct `MockConsumer`. The interface will be `MockConsumer`, and struct will be unexported `mockConsumer`

Reply:
I expanded it with all the methods we need:

```
type MockConsumer interface {
	ConsumeLogs(_ context.Context, ld plog.Logs) error
	ConsumeMetrics(_ context.Context, md pmetric.Metrics) error
	ConsumeTraces(_ context.Context, td ptrace.Traces) error
	Capabilities() consumer.Capabilities
	Clear()
	RequestCounter() RequestCounter
}
```

and I use it in `mockOtlpReceiver` like this:

```
type mockOtlpReceiver struct {
	mockConsumer    exportertest.MockConsumer
	traceReceiver   receiver.Traces
	metricsReceiver receiver.Metrics
	logReceiver     receiver.Logs
}
```